File: simlogiqueJSProto/js/core/Link.js
class Link {
	constructor(circuit, start, end) {
		this._circuit = circuit;
		this._start = start;
		this._end = end;
		this._connection = undefined;
		
		this._start._connect(this);
		this._end._connect(this);
		
		if(this._connection == undefined)
			this.setConnection(new Connection(this._circuit));
	}
	
	setConnection(connection) {
		if(connection != this._connection) {
			var oldConnection = this._connection;
			this._connection = connection;
			this._connection._addLink(this);
			this._start.setConnection(this._connection);
			this._end.setConnection(this._connection);
			if(oldConnection != undefined)
				oldConnection._removeLink(this);
		}
	}
	
	destructor() {
		this._start._disconnect(this);
		this._end._disconnect(this);
		if(this._start.isConnected() && this._end.isConnected())
			this._start.setConnection(new Connection(this._circuit));
		this._connection._removeLink(this);
	}
}

File: SimLogiqueCpp/Link.cpp
#include "Link.h"

// Link
Link::Link(Circuit& circuit, Connector& start, Connector& end) : _circuit(circuit), _start(start), _end(end) {
    _start.connect(*this);
    _end.connect(*this);

    if(_connection == nullptr)
        setConnection(new Connection(_circuit));
}

Link::~Link() {
    _start.disconnect(*this);
    _end.disconnect(*this);
    if(_start.isConnected() && _end.isConnected())
        _start.setConnection(new Connection(_circuit));
    _connection->removeLink(*this);
}

void Link::setConnection(Connection* connection) {
    if(connection != _connection) {
        Connection* oldConnection = _connection;
        _connection = connection;
        _connection->addLink(*this);
        _start.setConnection(_connection);
        _end.setConnection(_connection);
        if(oldConnection != nullptr)
            oldConnection->removeLink(*this);
    }
}
